internal/subscriber/subscriber.go

```diff
@@ -249,12 +189,68 @@ func (s *Subscriber) acknowledge(req *AckMessage) {
 		logger.Ctx(ctx).Infow("subscriber: max committed offset new value", "logFields", logFields, "offsetToCommit", offsetToCommit, "topic-partition", tp)
 	}
 
-	s.removeMessageFromMemory(ctx, stats, req.MessageID)
+	s.removeMessageFromMemory(ctx, stats, msg.MessageID)
 
 	// add to eviction map only in case of any out of order eviction
 	if offsetToCommit > stats.maxCommittedOffset {
 		stats.evictedButNotCommittedOffsets[offsetToCommit] = true
 	}
+}
+
+// acknowledge messages
+func (s *Subscriber) acknowledge(req *AckMessage) {
+	span, ctx := opentracing.StartSpanFromContext(req.ctx, "Subscriber:Ack", opentracing.Tags{
+		"subscriber":   req.SubscriberID,
+		"topic":        req.Topic,
+		"subscription": s.subscription.Name,
+		"message_id":   req.MessageID,
+		"partition":    req.Partition,
+	})
+	defer span.Finish()
+
+	if req == nil {
+		return
+	}
+
+	logFields := s.getLogFields()
+	logFields["messageId"] = req.MessageID
+
+	ackStartTime := time.Now()
+	defer func() {
+		logFields["ackTimeTaken"] = time.Now().Sub(ackStartTime).Seconds()
+		logger.Ctx(ctx).Infow("subscriber: ack request end", "logFields", logFields)
+	}()
+
+	logger.Ctx(ctx).Infow("subscriber: got ack request", "logFields", logFields)
+
+	tp := req.ToTopicPartition()
+	stats := s.consumedMessageStats[tp]
+
+	if stats.offsetBasedMinHeap.IsEmpty() {
+		return
+	}
+
+	msgID := req.MessageID
+	// check if message is present in-memory or not
+	if _, ok := stats.consumedMessages[msgID]; !ok {
+		logger.Ctx(ctx).Infow("subscriber: skipping ack as message not found in-memory", "logFields", logFields)
+		return
+	}
+
+	msg := stats.consumedMessages[msgID].(messagebroker.ReceivedMessage)
+
+	// if somehow an ack request comes for a message that has met deadline eviction threshold
+	if req.HasHitDeadline() {
```

Comment:
shouldn't this validation check happen outside of subscriber acknowledge, probably at controller layer ?

even if we need the check, i guess we should throw error here and let the caller handle it whether to retry or not

Reply:
So this check exists to check if a message has exceeded deadline and not been evicted by our ticker. In this case we throw it for retry. 

Currently the subscriber is the only entity which makes decisions on messages. I guess that is the reason why this is not modelled as you've said. 

Also we don't throw errors in case of acks and nacks today. Perhaps the reason is pubsub does not do it that way in case of streaming requests?